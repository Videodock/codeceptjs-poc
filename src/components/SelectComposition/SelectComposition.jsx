import React  from 'react';
import * as PropTypes from 'prop-types';
import { Box, Button, Container, IconButton, InputAdornment, TextField, Typography } from '@material-ui/core';
import { makeStyles } from '@material-ui/styles';
import Cancel from "@material-ui/icons/Cancel";
import { useTranslation } from 'react-i18next';
import images from '../../theme/images';
import styles from './SelectComposition.styles';

const useStyles = makeStyles(styles);

export default function SelectComposition({ composition, score, onSelectCompositionClick, onDeselectCompositionClick, onSelectScoreClick, onDeselectScoreClick, onBackButtonClick, onNextButtonClick }) {
  const { t }   = useTranslation('selectComposition');
  const classes = useStyles();

  return (
    <Box className={classes.root}>
      <img className={classes.person} src={images.personWithBass} alt={t('person')} />
      <Container className={classes.main}>
        <Typography variant="h2">{t('title')}</Typography>
        <Box className={classes.part}>
          <Typography variant="body1">{t('start')}</Typography>
          <Box className={classes.inputBox}>
            <TextField 
              value={composition?.name || t('composition_default')}
              variant="filled" 
              disabled={!composition}
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">
                    {!composition && <Button variant="contained" color="primary" onClick={onSelectCompositionClick}>{t('select')}</Button>}
                    {composition && <IconButton onClick={onDeselectCompositionClick}><Cancel /></IconButton> }
                  </InputAdornment>
                ) }
              }
              onChange={() => {}}
              fullWidth 
            />
          </Box>
          <Typography variant="body2">{t('note_compositions')}</Typography>
        </Box>
        <Box className={classes.part}>
          <Typography variant="body1">{t('score')}</Typography>
          <Box className={classes.inputBox}>
            <TextField 
              value={score? (score.title || score.name) : t('score_default')}
              // placeholder={t('score_default')}
              variant="filled" 
              disabled={!score}
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">
                    {!score && <Button variant="contained" color="primary" disabled={!composition} onClick={onSelectScoreClick}>{t('select')}</Button>}
                    {score && <IconButton onClick={onDeselectScoreClick}><Cancel /></IconButton> }
                  </InputAdornment>
                ) }
              }
              fullWidth 
            />
          </Box>
        </Box>
        <Box className={classes.formNav}>
          <Button onClick={onBackButtonClick}>{t('back')}</Button>
          <Button variant="contained" color="primary" disabled={!(composition && score) } onClick={onNextButtonClick}>{t('next')}</Button>
        </Box>
      </Container>
      <img className={classes.person} src={images.personTrumpetStandingRight} alt={t('person')} />
    </Box>
  );
}

SelectComposition.propTypes = { 
  composition               : PropTypes.object,
  score                     : PropTypes.object, 
  onSelectCompositionClick  : PropTypes.func, 
  onDeselectCompositionClick: PropTypes.func, 
  onSelectScoreClick        : PropTypes.func,
  onDeselectScoreClick      : PropTypes.func,
}; 