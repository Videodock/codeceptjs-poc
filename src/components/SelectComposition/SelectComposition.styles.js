import { createStyles } from '@material-ui/styles';

export default ({ spacing, palette, shape, breakpoints, typography }) => createStyles({
  root: {
    display      : 'flex',
    flexDirection: 'row',
  },
  main: {
    display       : 'flex',
    flexDirection : 'column',
    justifyContent: 'center',
    alignItems    : 'center',
    width         : '45vw',
  },
  formNav: {
    display       : 'flex',
    width         : '100%',
    justifyContent: 'flex-end',
  },
  person: {
    [breakpoints.down('sm')]: {
      display: 'none',
    },
    height: '50vh',
  },
  part: {
    width  : '100%',
    padding: spacing(3),
  },
  inputBox: {
    display      : 'flex',
    flexDirection: 'row',
    width        : '100%',
  },
});
