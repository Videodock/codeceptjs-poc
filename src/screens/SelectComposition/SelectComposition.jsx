import React, { useState } from "react";
import * as PropTypes from 'prop-types';
import { useTranslation } from "react-i18next";
import { useLazyQuery } from "@apollo/react-hooks";
import { gql } from "apollo-boost";
import { Box, Button, Container, Dialog, makeStyles, Typography } from "@material-ui/core";
import MultiModalComponent, { SearchConfig, searchTypes } from "trompa-multimodal-component";
import SelectCompositionComponent from "../../components/SelectComposition";
import SelectScoreModal from "../../components/SelectScoreModal";
import styles from './SelectComposition.styles';

const MODAL_NONE  = 0;
const MODAL_COMP  = 1;
const MODAL_SCORE = 2;

const searchConfig = new SearchConfig({
  searchTypes: [searchTypes.MusicComposition],
});

const useStyles = makeStyles(styles);

export default function SelectComposition({ onBackButtonClick, onCompositionSubmit, onSelectFileClick }) {
  const { t }                         = useTranslation('selectComposition');
  const classes                       = useStyles();
  const [composition, setComposition] = useState();
  const [score, setScore]             = useState();
  const [modal, setModal]             = useState(MODAL_NONE);

  const lazyQueryCallback                              = { onCompleted: data => onGqlLoaded(data), fetchPolicy: "no-cache" };
  const [getCompositionWithScores, { loading, error }] = useLazyQuery(GET_COMPOSITION_WITH_SCORES, lazyQueryCallback );

  /**
   * DATA 
   * Recieve node (composition) from MMC, then trigger lazyQuery to get metadata + scores
   * 
   * Todo: Get active campaigns of selected composition
   */
  const loadComposition = node => {
    setModal(MODAL_NONE);
    getCompositionWithScores({ variables: { identifier: node.identifier }, ...lazyQueryCallback });
  };
  const onGqlLoaded     = data => setComposition(data.MusicComposition?.[0]);
  
  /**
   * INTERNAL
   */
  const loadScore         = score => {
    setScore(score);
    setModal(MODAL_NONE);
  };
  const deselectBoth      = () => {
    setComposition();
    setScore();
  };
  const onNextButtonClick = () => onCompositionSubmit({ composition: composition, score: score });

  if (loading) return <Box>{t('loading')}</Box>;
  if (error) console.log('Failed to load scores', error);

  return (
    <Container className={classes.root}>
      <SelectCompositionComponent 
        composition={composition} 
        score={score}
        onSelectCompositionClick={() => setModal(MODAL_COMP)} 
        onDeselectCompositionClick={deselectBoth}
        onSelectScoreClick={() => setModal(MODAL_SCORE)}
        onDeselectScoreClick={() => setScore()} 
        onSelectFileClick={onSelectFileClick}
        onBackButtonClick={onBackButtonClick}
        onNextButtonClick={onNextButtonClick}
      />
      <Dialog
        className={classes.dialog}
        open={modal !== MODAL_NONE} 
        onClose={e => setModal(MODAL_NONE)}
        maxWidth="lg"
        fullWidth
      >
        <Box className={classes.dialogBox}>
          <Typography variant="h2">
            {modal === MODAL_COMP && t('select_composition')}
            {modal === MODAL_SCORE && t('select_score')}
          </Typography>
          <Button className={classes.closeButton} onClick={() => setModal(MODAL_NONE)}>X</Button>
          {modal === MODAL_COMP && (
            <MultiModalComponent 
              config={searchConfig}
              onResultClick={node => loadComposition(node)} 
            />
          )}
          {modal === MODAL_SCORE && (
            <SelectScoreModal 
              composition={composition}
              onLoadScore={node => loadScore(node)} 
            />
          )}
        </Box>
      </Dialog>
    </Container>
  );
}

SelectComposition.propTypes = {
  onBackButtonClick  : PropTypes.func, 
  onCompositionSubmit: PropTypes.func,
  onSelectFileClick  : PropTypes.func,
};

export const GET_COMPOSITION_WITH_SCORES = gql`
  query GetMusicCompositionWithScores($identifier: ID!){
      MusicComposition(identifier: $identifier) {
      __typename
      identifier
      name
      workExample {
        ... on DigitalDocument {
          identifier
          contributor
          creator
          description
          format
          image
          language
          name
          publisher
          relation
          rights
          source
          subject
          title
        }
      }
    }
  }
`;