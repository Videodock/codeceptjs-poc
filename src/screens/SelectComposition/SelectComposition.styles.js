import { createStyles } from '@material-ui/styles';

export default ({ spacing, palette }) => createStyles({
  root: {
    width  : '100vw',
    height : '100%',
    padding: spacing(5),
  },
  dialog: {
    backgroundColor: palette.background.primary,
  },
  dialogBox: {
    display       : 'flex',
    flexDirection : 'column',
    justifyContent: 'center',
    alignItems    : 'center',
    padding       : spacing(3),
  },
  closeButton: {
    position: 'absolute',
    right   : spacing(3),
    top     : spacing(3),
  },
});
