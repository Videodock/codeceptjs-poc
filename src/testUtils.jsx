import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { ThemeProvider } from '@material-ui/styles';
import { I18nextProvider } from 'react-i18next';
import { ApolloProvider } from '@apollo/client';
import { BrowserRouter as Router } from 'react-router-dom';
import 'url-search-params-polyfill';
import i18n from '../src/i18n';
import client from './graphql';
import theme from './theme';
import store from './redux';

const wrapper = ({ children }) => (
  <ApolloProvider client={client}>
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <I18nextProvider i18n={i18n}>
          <Router>
            {children}
          </Router>
        </I18nextProvider>
      </ThemeProvider>
    </Provider>
  </ApolloProvider>
);

const customRender = (ui, options) => render(ui, { wrapper, ...options });

export { customRender as render };
