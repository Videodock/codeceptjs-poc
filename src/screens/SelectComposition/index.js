import SelectComposition from "./SelectComposition";

export * from './SelectComposition';
export default SelectComposition;
