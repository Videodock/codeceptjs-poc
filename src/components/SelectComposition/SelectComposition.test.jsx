import React from 'react';
import { render } from '../../testUtils';
import SelectComposition from './SelectComposition';

describe('<SelectComposition />', () => {
  test('renders and matches snapshot', () => {
    const { container } = render(<SelectComposition />);

    expect(container.firstChild).toMatchSnapshot();
  });

  // add your unit tests here
});
