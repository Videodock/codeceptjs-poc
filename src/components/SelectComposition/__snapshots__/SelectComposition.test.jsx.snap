// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectComposition /> renders and matches snapshot 1`] = `
<div
  class="MuiBox-root MuiBox-root-8 makeStyles-root-1"
>
  <img
    alt="person"
    class="makeStyles-person-4"
    src="personWithBass.png"
  />
  <div
    class="MuiContainer-root makeStyles-main-2 MuiContainer-maxWidthLg"
  >
    <h2
      class="MuiTypography-root MuiTypography-h2"
    >
      Composition & Score
    </h2>
    <div
      class="MuiBox-root MuiBox-root-47 makeStyles-part-5"
    >
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        Let's start with the composition
      </p>
      <div
        class="MuiBox-root MuiBox-root-48 makeStyles-inputBox-6"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <div
            class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline Mui-disabled Mui-disabled MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiFilledInput-adornedEnd"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiFilledInput-input Mui-disabled Mui-disabled MuiInputBase-inputAdornedEnd MuiFilledInput-inputAdornedEnd"
              disabled=""
              placeholder="Music Composition"
              type="text"
              value=""
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-filled MuiInputAdornment-positionEnd"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Select
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
      <p
        class="MuiTypography-root MuiTypography-body2"
      >
        In Trompa we support a wide range of compositions
      </p>
    </div>
    <div
      class="MuiBox-root MuiBox-root-127 makeStyles-part-5"
    >
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        Score
      </p>
      <div
        class="MuiBox-root MuiBox-root-128 makeStyles-inputBox-6"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <div
            class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline Mui-disabled Mui-disabled MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiFilledInput-adornedEnd"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiFilledInput-input Mui-disabled Mui-disabled MuiInputBase-inputAdornedEnd MuiFilledInput-inputAdornedEnd"
              disabled=""
              placeholder="Score"
              type="text"
              value=""
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-filled MuiInputAdornment-positionEnd"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Select
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-129 makeStyles-formNav-3"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Back
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Next
        </span>
      </button>
    </div>
  </div>
  <img
    alt="person"
    class="makeStyles-person-4"
    src="personTrumpetStandingRight.png"
  />
</div>
`;
